engine/engine.go

```diff
@@ -256,7 +258,7 @@ func (e *Engine) RegisterMultihashLister(mhl provider.MultihashLister) {
 func (e *Engine) NotifyPut(ctx context.Context, contextID []byte, md metadata.Metadata) (cid.Cid, error) {
 	// The multihash lister must have been registered for the linkSystem to know how to
 	// go from contextID to list of CIDs.
-	return e.publishAdvForIndex(ctx, contextID, md, false)
+	return e.PublishAdvForIndex(ctx, contextID, nil, e.provider.ID, e.retrievalAddrsAsString(), md, false)
```

Comment:
Further to comments from @willscott please consider using the existing `Publish` API that takes `schema.Advertisement` which closely matches the HTTP request struct introduced in this PR and avoid exposing internal Engine APIs :)

Reply:
since publish a `advertisement` from http request is with the same logical of `publishAdvForIndex`,
we needs to manage the `metadata` of advertisements,  
so following methods in `Engine` are all needs by this logical:

https://github.com/filecoin-project/index-provider/blob/cf3e8e80afdc62bf0ce94cfcb375f73e751b50cb/engine/engine.go#L455-L510

i had successfully construct an `Advertisement` out `Engine`, and call `Publish`, this is ok.

but if do **NOT** expose any method,  you known, it is hard. no way to manage the `metadata`

i still think the good way is expose `publishAdvForIndex`.

hope your suggestion. thanks..

Reply:
> no way to manage the metadata

Can you expand on this please.

Reply:
ok, i may find another way to do this.
not expose `publishAdvForIndex` and have no change to `Engine`. will working on this.. (:-

Reply:
> > no way to manage the metadata
> 
> Can you expand on this please.

tomorrow, it's getting late in china now.

Reply:
> Can you expand on this please.

ok, i will try to explain the `hard`, 
assume we create a advertisement http server to handle advertisement publishing request,  and not expose any method in `engine`, 

this http server should be defined like this:
```go
Server struct {
      ....
	e  *engine.Engine
	k  crypto.PrivKey
	ce *chunker.CachedEntriesChunker //
	ds datastore.Batching
}
```

just a **not remove** advertisement for example, the logical of handler should be like:
```go
func (s *Server) advertisementHandler(w http.ResponseWriter, r *http.Request) {
	var (
		req AnnounceAdvReq
		err error
	)
       ......
	ctx := r.Context()
	
	contextCID, _ := s.ds.Get(cidMapPrefix+req.ContextID)
	if contextCID == cid.Undef {
		lnk := ... generate link list ...
		s.ds.Put(ctx, cidMapPrefxi+req.ContextID, lnk.Cid)
	} else {
		prevMetadata, _ := e.ds.Get(keyMetadataMapPrefix+req.ContextID)
		if req.Md.Equal(prevMetadata) {
			... response error ...
		}
	}
	s.ds.Put(ctx, keyToMetadataMapPrefix+req.ContextID, &req.Md)

	adv, err := s.makeAdvertisement(ctx, ...)
	if err != nil {
		... resposne error ...
	}
	adv.Sign(s.k)
	
	// finally we can call `engine.Publish`
	s.e.Publish(ctx, adv)
	
	... respond result ...
}
```

what the `Server.advertisementHandler` do is an exactly copy of `Engine.PublishAdvForIndex`.

i said hard to mange `metadata`, means those  `e.ds.Get`, `e.ds.Put` in above code,  are another exactly copy of operations of the badger store , 
witch is already implemented in `Engine`

Reply:
Re Context ID and metadata determination in this code snippet:
```go
contextCID, _ := s.ds.Get(cidMapPrefix+req.ContextID)
	if contextCID == cid.Undef {
		lnk := ... generate link list ...
		s.ds.Put(ctx, cidMapPrefxi+req.ContextID, lnk.Cid)
	} else {
		prevMetadata, _ := e.ds.Get(keyMetadataMapPrefix+req.ContextID)
		if req.Md.Equal(prevMetadata) {
			... response error ...
		}
	}
	s.ds.Put(ctx, keyToMetadataMapPrefix+req.ContextID, &req.Md)
```

This is _entirely application specific_. The point of libraries offered by this repository is to be application agnostic. That logic cannot live in this repo. Because, the way Context ID or Metadata determined by different application is different. And that's why they are of type `[]byte` in the advertisement schema.

An alternative approach to make the implementation of the wrapper `Server` on your side simpler is to use `provider.MultihashLister` and [register](https://github.com/filecoin-project/index-provider/blob/main/interface.go#L40) an implementation of it to the `Engine`. That would then allow the logic by which a `contextID` can retrieve the list of multihashes for a an advertisement to be customised. The server can then use `NotifyPut` and `NotifyRemove` to publish advertisements instead of calling `Publish`.

Reply:
This is a good idea
I had tried this solution yesterday, the reason i gave up was i didn't kwon how to pass `ProviderID`, `Retrieval address` to the `Engine`.
and you just said:
> ProvideID can be passed today to the Engine by specifying libp2p [host](https://github.com/filecoin-project/index-provider/blob/main/engine/options.go#L224). Retrieval address can also be specified by [this](https://github.com/filecoin-project/index-provider/blob/main/engine/options.go#L245) option

I m a little confused about this,  when the `daemon` create the `engine` instance:
https://github.com/filecoin-project/index-provider/blob/cf3e8e80afdc62bf0ce94cfcb375f73e751b50cb/cmd/provider/daemon.go#L119-L126
the `host` of index provider is created here, according to your above suggestion, does it means:
I should  create another **provider** `host`,  may be need add a configuration for this? 
like the following fake code:

```go
var providerHost = h
if provider_configuration_exists {
   providerHot = createProviderHost()
}
// Starting provider core
eng, err := engine.New(
...
	engine.WithHost(providerHost),
...

```

thanks for your patience @masih

Reply:
No worries; happy to be of help.

The `daemon` is meant to be a service that publishes CAR files only. A bit of history: it was written as a tool for internal testing when network indexer was being developed. 

As it stands today, `daemon` reads that information from a config file. So you can customise those values in the config file, and set any options not set from the config when `daemon` is instantiated. But keep in mind that `daemon` is not designed for dynamically changing retrieval address etc.

A better approach here may be to implement a separate HTTP server that takes an `engine` instance, then optionally expose that via `daemon` as a dedicated server. If that code is going to be application specific, then my recommendation is to move that to a repo close to where Venus market is located, and try not to reuse `daemon`. Instead start up that wrapper directly in a `main` function.

Reply:
Tanks a lot. I think i get it. .....LOL...
I should not think of the index-provider to be an app, it should be a library. and the cmd/provider is just a testing tool.
I had been getting stuck in the thinking of adding this functionary on this testing tool.